File: packages/mcp-authz/src/verifier.ts
import {
  OAuthError,
  OAuthErrorCode,
  type AuthInfo,
  type OAuthTokenVerifier,
} from '@modelcontextprotocol/server';
import { createRemoteJWKSet, jwtVerify } from 'jose';
import type { Identity } from './identity';

/**
 * Token verification, the one thing the SDK deliberately leaves to you.
 *
 * `authInfo` is strictly pass-through in the SDK: it is never derived from
 * request headers, and no token is checked unless we check it. So this file is
 * the security boundary of the whole server.
 *
 * The authorization server is something that already exists (WorkOS, Stytch,
 * Auth0) doing dynamic client registration / CIMD and Google Workspace login.
 * Google cannot play that role itself: it has no DCR/CIMD, and it will not
 * mint a token whose audience is this server.
 */

export type VerifierOptions = {
  /** The AS issuer, e.g. `https://auth.acme.com`. Must match the token's `iss`. */
  issuer: string;
  /** Where the AS publishes its signing keys. */
  jwksUri: string;
  /**
   * This server's public URL. A token minted for a different resource is
   * refused even when its signature is valid (RFC 8707).
   */
  resource: URL;
  /**
   * Restrict to one Google Workspace domain, checked against the `hd` claim
   * the AS passes through. Omit to accept any domain the AS admits.
   */
  allowedDomain?: string;
  /** Claim carrying the verified email. Auth0 and WorkOS both use `email`. */
  emailClaim?: string;
  /** Claim proving the email was verified. Defaults to `email_verified`. */
  emailVerifiedClaim?: string;
  /** Require an explicit `true` verified-email claim. Defaults to true. */
  requireEmailVerified?: boolean;
};

const invalidToken = (message: string) => new OAuthError(OAuthErrorCode.InvalidToken, message);

/**
 * A JWKS-backed verifier. Keys are fetched once and cached by `jose`, which
 * also handles rotation, so a key roll at the AS does not need a redeploy.
 */
export function jwksVerifier(options: VerifierOptions): OAuthTokenVerifier & {
  identityOf: (auth: AuthInfo) => Identity;
} {
  const jwks = createRemoteJWKSet(new URL(options.jwksUri));
  const emailClaim = options.emailClaim ?? 'email';
  const emailVerifiedClaim = options.emailVerifiedClaim ?? 'email_verified';
  // Compared without the hash fragment, per the AuthInfo.resource contract.
  const expectedResource = new URL(options.resource.href).href.split('#')[0]!;

  return {
    async verifyAccessToken(token: string): Promise<AuthInfo> {
      let payload: Record<string, unknown>;
      try {
        // `audience` is the resource-indicator check. Skipping it is the bug
        // that lets a token minted for another service be replayed here.
        const verified = await jwtVerify(token, jwks, {
          issuer: options.issuer,
          audience: expectedResource,
        });
        payload = verified.payload as Record<string, unknown>;
      } catch (error) {
        throw invalidToken(`Token rejected: ${error instanceof Error ? error.message : String(error)}`);
      }

      if (typeof payload.exp !== 'number') {
        throw invalidToken('Token has no `exp` claim.');
      }

      const email = payload[emailClaim];
      if (typeof email !== 'string' || !email) {
        throw invalidToken(`Token carries no '${emailClaim}' claim, so there is no identity to map.`);
      }
      if ((options.requireEmailVerified ?? true) && payload[emailVerifiedClaim] !== true) {
        throw invalidToken(`Token does not prove '${emailClaim}' with '${emailVerifiedClaim}: true'.`);
      }

      const sub = payload.sub;
      if (typeof sub !== 'string' || !sub) {
        throw invalidToken('Token has no `sub` claim, so there is no stable subject to bind to.');
      }

      const domain = typeof payload.hd === 'string' ? payload.hd : undefined;
      if (options.allowedDomain && domain?.toLowerCase() !== options.allowedDomain.toLowerCase()) {
        throw invalidToken(`Token is for ${domain ?? 'an unknown domain'}, not ${options.allowedDomain}.`);
      }

      return {
        token,
        clientId:
          typeof payload.client_id === 'string'
            ? payload.client_id
            : typeof payload.azp === 'string'
              ? payload.azp
              : sub,
        scopes: scopesOf(payload.scope),
        expiresAt: payload.exp,
        resource: options.resource,
        // The whole payload rides along so policy can match on IdP group claims
        // without this file having to know which claim an IdP puts them in.
        extra: {
          issuer: options.issuer,
          sub,
          email,
          emailVerified: options.requireEmailVerified === false || payload[emailVerifiedClaim] === true,
          domain,
          claims: payload,
        },
      };
    },

    identityOf: identityFromAuth,
  };
}

/** Default identity mapper for custom verifiers using `AuthInfo.extra`. */
export function identityFromAuth(auth: AuthInfo): Identity {
  const { issuer, sub, email, emailVerified, domain, claims } = auth.extra ?? {};
  if (typeof issuer !== 'string' || !issuer || typeof sub !== 'string' || !sub) {
    throw invalidToken('Verified token carried no issuer or subject.');
  }
  if (email !== undefined && (typeof email !== 'string' || !email)) {
    throw invalidToken('Verified token carried an invalid email.');
  }
  return {
    issuer,
    sub,
    email: typeof email === 'string' ? email : undefined,
    emailVerified: emailVerified === true,
    domain: typeof domain === 'string' ? domain : undefined,
    claims: isRecord(claims) ? claims : {},
  };
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

/** OAuth scope is a space-delimited string; some servers send an array anyway. */
function scopesOf(scope: unknown): string[] {
  if (Array.isArray(scope)) return scope.filter((s): s is string => typeof s === 'string');
  if (typeof scope === 'string') return scope.split(' ').filter(Boolean);
  return [];
}

File: packages/mcp-authz-python/src/mcp_authz/verifier.py
"""JWKS-backed verification using the official MCP TokenVerifier interface."""

from __future__ import annotations

import asyncio
from collections.abc import Sequence
from typing import Any

import jwt
from jwt import PyJWKClient
from jwt.exceptions import PyJWTError
from mcp.server.auth.provider import AccessToken


class JwtVerifier:
    """Verify signed JWT access tokens into official MCP ``AccessToken`` values."""

    def __init__(
        self,
        *,
        issuer: str,
        jwks_uri: str,
        resource: str,
        allowed_domain: str | None = None,
        email_claim: str = "email",
        email_verified_claim: str = "email_verified",
        require_email_verified: bool = True,
        algorithms: Sequence[str] = ("RS256", "ES256"),
    ) -> None:
        self.issuer = issuer
        self.resource = resource.split("#", maxsplit=1)[0]
        self.allowed_domain = allowed_domain
        self.email_claim = email_claim
        self.email_verified_claim = email_verified_claim
        self.require_email_verified = require_email_verified
        self.algorithms = tuple(algorithms)
        if not self.algorithms or any(algorithm.startswith("HS") for algorithm in self.algorithms):
            raise ValueError("JwtVerifier requires at least one asymmetric signing algorithm.")
        self._jwks = PyJWKClient(jwks_uri, cache_keys=True)

    async def verify_token(self, token: str) -> AccessToken | None:
        """Return ``None`` for every invalid credential, as the MCP SDK expects."""

        return await asyncio.to_thread(self._verify_token, token)

    def _verify_token(self, token: str) -> AccessToken | None:
        try:
            key = self._jwks.get_signing_key_from_jwt(token)
            payload = jwt.decode(
                token,
                key.key,
                algorithms=list(self.algorithms),
                audience=self.resource,
                issuer=self.issuer,
                options={"require": ["exp", "iss", "aud", "sub"]},
            )
        except (PyJWTError, ValueError):
            return None

        sub = payload.get("sub")
        email = payload.get(self.email_claim)
        if not isinstance(sub, str) or not sub or not isinstance(email, str) or not email:
            return None
        if self.require_email_verified and payload.get(self.email_verified_claim) is not True:
            return None
        domain = payload.get("hd")
        if self.allowed_domain and (not isinstance(domain, str) or domain.casefold() != self.allowed_domain.casefold()):
            return None

        normalized_claims = dict(payload)
        normalized_claims["email"] = email
        normalized_claims["email_verified"] = (
            not self.require_email_verified or payload.get(self.email_verified_claim) is True
        )
        if isinstance(domain, str):
            normalized_claims["hd"] = domain
        return AccessToken(
            token=token,
            client_id=_client_id(payload, sub),
            scopes=_scopes(payload.get("scope")),
            expires_at=int(payload["exp"]),
            resource=self.resource,
            subject=sub,
            claims=normalized_claims,
        )


def _client_id(payload: dict[str, Any], fallback: str) -> str:
    for claim in ("client_id", "azp"):
        value = payload.get(claim)
        if isinstance(value, str) and value:
            return value
    return fallback


def _scopes(value: Any) -> list[str]:
    if isinstance(value, str):
        return value.split()
    if isinstance(value, list):
        return [scope for scope in value if isinstance(scope, str) and scope]
    return []
